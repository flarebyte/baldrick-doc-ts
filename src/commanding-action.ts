import { updateAll } from './io-sfx.js';
import { GenerateTypedocAction, GenerateTypedocActionOpts, RunnerContext } from './model.js';

export const cmdGenerateTypedocAction: GenerateTypedocAction = async (
  ctx: RunnerContext,
  options: GenerateTypedocActionOpts
) => {
  await updateAll(ctx, options);
  ctx.termFormatter({
    title: 'Generated typedoc documentation',
    detail: [
      'Check stuff',
    ].join('\n'),
    kind: 'info',
    format: 'default',
  });
};