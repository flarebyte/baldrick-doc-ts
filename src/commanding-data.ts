import {
	type CmdOption,
	type CmdOptionsGeneratorTypedoc,
	type CmdOptionsParser,
} from './model.js';

const feature: CmdOption = {
	shortFlag: 'f',
	longFlag: 'feature',
	description: 'List of features',
	choices: ['md'],
	mandatory: false,
	variadic: true,
	defaultValue: ['md'],
};

const jsonSource: CmdOption = {
	shortFlag: 's',
	longFlag: 'json-source',
	description: 'Typedoc json filename',
	defaultValue: 'doc.json',
	choices: [],
	mandatory: false,
	variadic: false,
};
const docBase: CmdOption = {
	shortFlag: 'db',
	longFlag: 'doc-base',
	description: 'Specify the base name for documentation (doc/api)',
	defaultValue: '',
	choices: [],
	mandatory: false,
	variadic: false,
};
const srcDirectory: CmdOption = {
	shortFlag: 'sd',
	longFlag: 'src-directory',
	description: 'Directory of the source code that will be used for links',
	choices: [],
	mandatory: true,
	variadic: false,
	defaultValue: 'src',
};

export const cmdOptionsGenerator: CmdOptionsGeneratorTypedoc = {
	feature,
	jsonSource,
	docBase,
	srcDirectory,
};

const featureParser: CmdOption = {
	shortFlag: 'f',
	longFlag: 'feature',
	description: 'List of features',
	choices: ['internal', 'functions-csv', 'ngram'],
	mandatory: false,
	variadic: true,
	defaultValue: ['internal', 'ngram'],
};

export const cmdOptionsParser: CmdOptionsParser = {
	feature: featureParser,
	docBase,
	srcDirectory,
};
