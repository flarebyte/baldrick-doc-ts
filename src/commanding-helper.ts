import path from 'node:path';
import {Option} from 'commander';
import {type CmdOption} from './model.js';

const capitalize = (value: string): string =>
	value.length > 0 ? (value[0] || '').toUpperCase() + value.slice(1) : '';

const decapitalize = (value: string): string =>
	value.length > 0 ? (value[0] || '').toLowerCase() + value.slice(1) : '';

export const toCamelCase = (longFlag: string): string =>
	decapitalize(longFlag.split('-').map(capitalize).join(''));

export const toCommanderOption = (option: CmdOption): Option => {
	const dot3 = option.variadic ? '...' : '';
	const flags = `-${option.shortFlag}, --${option.longFlag} [${toCamelCase(
		option.longFlag,
	)}${dot3}]`;
	const options = new Option(flags, option.description);
	options.defaultValue = option.defaultValue;
	if (option.choices.length > 0) {
		options.choices(option.choices);
	}

	options.mandatory = option.mandatory;
	return options;
};

export const splitDocBase = (
	base: string,
): {docDirectory: string; docPrefix: string} =>
	base === ''
		? {docDirectory: '', docPrefix: ''}
		: {
			docDirectory: path.dirname(base),
			docPrefix: path.basename(base),
		};
