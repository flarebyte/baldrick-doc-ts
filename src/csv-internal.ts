import CSV from 'papaparse';
import { FunctionInfo, ModuleInfo } from './parser-model.js';

interface CsvInternalRow {
  filename: string;
  functionName: string;
  descendantCount: number;
  bodyWidth: number;
  keywords: string;
}

const getKeywords = (functionInfo: FunctionInfo): string[] => {
  const isExpression = functionInfo.expression ? ['expression']: [];
  const isExported = functionInfo.exported ? ['public']: ['private']
  return [...isExported, ...isExpression ]
}

const fromFunctionInfo =
  (filename: string) =>
  (functionInfo: FunctionInfo): CsvInternalRow => ({
    filename,
    functionName: functionInfo.identifier,
    descendantCount: functionInfo.descendantCount,
    bodyWidth: functionInfo.bodyWidth,
    keywords: getKeywords(functionInfo).join(' ')
  });

const sortedByFunction = (
  a: { functionName: string },
  b: { functionName: string }
): number => {
  if (a.functionName > b.functionName) return 1;
  if (a.functionName < b.functionName) return -1;
  return 0;
};

export const toCsvFonctions = (module: ModuleInfo): string => {
  const internalFunctions = module.sources
    .flatMap((source) =>
      source.functions.map(fromFunctionInfo(source.filename))
    )
    .sort(sortedByFunction);
  return CSV.unparse(internalFunctions, {
    columns: ['functionName', 'bodyWidth', 'descendantCount', 'keywords' ,'filename'],
  });
};
