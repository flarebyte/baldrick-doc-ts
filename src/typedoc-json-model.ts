export type TypedocJson = {
	name: string;
	originalName: string;
	children: TypedocChild[];
	groups: Group[];
};

export type TypedocChild = {
	id: number;
	name: string;
	kindString: string;
	flags: ChildFlags;
	sources: Source[];
	type?: Type;
	defaultValue?: string;
	signatures?: Signature[];
	comment?: Comment;
};

type ChildFlags = {
	isConst?: boolean;
};

type Source = {
	fileName: string;
	line: number;
	character: number;
};

type Type = {
	type: string;
	name?: string;
	elementType?: ElementType;
	declaration?: Declaration;
	value?: boolean;
};

type ElementType = {
	type: string;
	name: string;
};

type Signature = {
	id: number;
	name: string;
	kind: number;
	kindString: string;
	parameters: Parameter[];
	type: Type;
	comment?: Comment;
};

export type Parameter = {
	id: number;
	name: string;
	kind: number;
	kindString: string;
	type: Type;
	comment?: Comment;
	defaultValue?: string;
};

type Declaration = {
	id: number;
	name: string;
	kind: number;
	kindString: string;
	signatures: Signature[];
};

type Comment = {
	shortText: string;
	returns: string;
};

type Group = {
	title: string;
	kind: number;
	children: number[];
};
