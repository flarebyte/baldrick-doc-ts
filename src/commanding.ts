import { Command } from 'commander';
import { cmdOptionsGenerator } from './commanding-data.js';
import { splitDocBase, toCommanderOption } from './commanding-helper.js';
import { toFeatures } from './feature-helper.js';
import {
  GenerateTypedocAction,
  GenerateTypedocActionOpts,
  GenerateTypedocRawOpts,
  RunnerContext,
} from './model.js';
import { basicFormatter, errorFormatter } from './term-formatter.js';
import { version } from './version.js';

export class Commanding {
  _program: Command = new Command();
  constructor() {
    this._program.version(version);
  }

  declareGenerateAction(genAction: GenerateTypedocAction) {
    this._program
      .command('generate')
      .description('Generate and standardize the source code files')
      .addOption(toCommanderOption(cmdOptionsGenerator.feature))
      .addOption(toCommanderOption(cmdOptionsGenerator.jsonSource))
      .addOption(toCommanderOption(cmdOptionsGenerator.docBase))
     
      .action(async (options: GenerateTypedocRawOpts) => {
        const {
          feature,
          jsonSource,
          docBase,
        } = options;
        const generateTypedocOpts: GenerateTypedocActionOpts = {
          feature: toFeatures(feature),
          jsonSource,
          docBase,
          ...splitDocBase(docBase)
        };
        const ctx: RunnerContext = {
          currentPath: process.cwd(),
          termFormatter: basicFormatter,
          errTermFormatter: errorFormatter,
        };
        await genAction(ctx, generateTypedocOpts);
      });
  }

  async parseAsync(argv: string[]) {
    return await this._program.parseAsync(argv, { from: 'node' });
  }

  async parseAsyncArgv() {
    return await this.parseAsync(process.argv);
  }
}