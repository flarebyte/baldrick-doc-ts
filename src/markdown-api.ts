import { markdownToString } from './markdown.js';
import { MdSection } from './model.js';
import { Parameter, TypedocChild, TypedocJson } from './typedoc-json-model.js';
const bq = '`';

const parameterToString = (param: Parameter): string => {
  const { name, type, comment } = param;
  const description = comment?.shortText || 'fixme: Adds a description';

  const isArray = type.type === 'array';
  const arrayMarker = isArray ? 'Array of ' : '';
  const typeName = type.name || type.elementType?.name || 'fixme';
  return `* ${name}: ${arrayMarker}${bq}${typeName}${bq}: ${description}`;
};

const functionToMdSection = (child: TypedocChild): MdSection => {
  const sourceFilename = child.sources[0]?.fileName || 'which-file.ts';
  const sourceFileLine = child.sources[0]?.line || 1;
  const fileRef = `See [${sourceFilename} - L${sourceFileLine}](src/${sourceFilename})`;
  const signatures = child.signatures || [];
  const [signature] = signatures;
  const params = (signature?.parameters || []).map(parameterToString);
  const description =
    signature?.comment?.shortText || 'fixme: Adds a description';
  const returns = signature?.comment?.returns;
  const functionReturn = returns ? ['### Returns', returns] : [];

  const section: MdSection = {
    title: `${child.name}`,
    body: [
      description,
      '### Parameters',
      ...params,
      ...functionReturn,
      fileRef,
    ].join('\n\n'),
  };
  return section;
};

const variableToMdSection = (child: TypedocChild): MdSection => {
  const sourceFilename = child.sources[0]?.fileName || 'which-file.ts';
  const sourceFileLine = child.sources[0]?.line || 1;
  const fileRef = `See [${sourceFilename} - L${sourceFileLine}](src/${sourceFilename})`;
  const description = child?.comment?.shortText || 'fixme: Adds a description';
  const isArray = child?.type?.type === 'array';
  const arrayMarker = isArray ? 'Array of ' : '';

  const name = child?.type?.elementType?.name;
  const signature = `${arrayMarker}${bq}${name}${bq}`;

  const section: MdSection = {
    title: `${child.name}`,
    body: [signature, description, fileRef].join('\n\n'),
  };
  return section;
};

const titleToRef = (section: MdSection): string =>
  `* [${section.title}](#${section.title})`;

export const toTypedocApiMd = (typedocJson: TypedocJson): string => {
  const allFunctions = typedocJson.children.filter(
    (child) => child.kindString === 'Function'
  );
  const allVariables = typedocJson.children.filter(
    (child) => child.kindString === 'Variable'
  );
  const functionSections = allFunctions.map(functionToMdSection);
  const varSections = allVariables.map(variableToMdSection);
  const idxFunctions = functionSections.map(titleToRef);
  const idxVars = varSections.map(titleToRef);
  const mainSectionFunctions =
    idxFunctions.length === 0
      ? []
      : ['__Functions:__', '', ...idxFunctions, ''];
  const mainSectionVariables =
    idxVars.length === 0 ? [] : ['', '__Variables:__', '', ...idxVars];

  const mainSection = [...mainSectionFunctions, ...mainSectionVariables].join(
    '\n'
  );

  const mdDoc = {
    title: `API of ${typedocJson.name}`,
    description: `List of functions and variables for ${bq}${typedocJson.name}${bq}`,
    mainSection,
    sections: [...functionSections, ...varSections],
  };
  return markdownToString(mdDoc);
};
