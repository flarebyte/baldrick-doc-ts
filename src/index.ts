import {
	cmdGenerateTypedocAction,
	cmdParseAction,
} from './commanding-action.js';
import {Commanding} from './commanding.js';

const commanding = new Commanding();

commanding.declareGenerateAction(cmdGenerateTypedocAction);
commanding.declareParseAction(cmdParseAction);

export {commanding};
