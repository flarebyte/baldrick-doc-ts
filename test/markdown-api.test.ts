import { toTypedocApiMd } from '../src/markdown-api';
import { TypedocJson } from '../src/typedoc-json-model';
import { typedocExampleJson } from './fixture-typedoc';

describe('markdown-api', () => {
  const typedocExample: TypedocJson = JSON.parse(typedocExampleJson);
  it('should provide', () => {
    const actual = toTypedocApiMd(typedocExample);
    expect(actual).toMatchSnapshot();
  });
});
