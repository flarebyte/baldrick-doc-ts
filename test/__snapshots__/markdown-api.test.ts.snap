// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`markdown-api should provide 1`] = `
"# API of object-crumble

> List of functions and variables for \`object-crumble\`

__Functions:__

* [abstractObject](#abstractObject)
* [anyOfString](#anyOfString)
* [mutateObject](#mutateObject)
* [someUrl](#someUrl)


__Variables:__

* [mutatorRules](#mutatorRules)

## abstractObject

Convert any object to an abstract representation of the object structure
\`\`\`
abstractObject({name: \\"Jane\\"})
\`\`\`

### Parameters

* rules: Array of \`StringAbstractionRule\`: a list of rules used to infer the kind of value of each property in the object

* prefix: \`string\`: path of the object in the enclosing object, or empty if none

### Returns

an array of \`CrumbleAbstractedValue\` representing each \`path\` \`kind\` pair.


See [obj-abstractor.ts - L58](src/obj-abstractor.ts)

## anyOfString

Detect a a string from a given list of string

### Parameters

* name: \`string\`: the keyword to return if the detection happens

* options: Array of \`string\`: a list of string that would satisfy the detection

### Returns

the \`name\` keyword or false


See [string-abstractor.ts - L19](src/string-abstractor.ts)

## mutateObject

Mutates an object applying a list of mutation.
We are using currying to provide the list of mutations rules.
Only a single mutation is applied to facilitate the studying of the impact.

### Parameters

* rules: Array of \`CrumbleFieldMutation\`: a list of rules used to infer the kind of value of each property in the object

### Returns

a \`CrumbleObject\` representing most javascript objects


See [obj-mutator.ts - L104](src/obj-mutator.ts)

## someUrl

Detect an url

### Parameters

* value: \`string\`: the text to check

### Returns

the \`url\` keyword or false


See [string-abstractor.ts - L9](src/string-abstractor.ts)

## mutatorRules

Array of \`CrumbleFieldMutation\`

Basic mutator rules that can be used out of the box

See [obj-mutator.ts - L38](src/obj-mutator.ts)"
`;
